Refactor CDC Debezium Source

tzolov opened the pull request:
- New `debezium-source` implementation, build on the Debezium Engine API. No forking of Debezium code.
 - Based on latest `2.2.0 CR1` Debezium version.
 - Supports for `MySQL`, `PostgreSQL`, `Oracle`, ` SQL Server`, `MongoDB`, `DB2`, `Vitess` and `Spanner` datastores.
 - Supports `JSON`, `AVRO` and (Experimental) `PROTOBUF` content formats with `byte[]` encoding. Defaults to `JSON`.
 - Configuration is split into Debezium Engine auto-config and Consumer change-event configuration. Later allows for plugging different consumer implementation.
    - Default Consumer implementation obeys the functional Supplier contract (e.g. exposes a `Supplier<Flux<Message<?>>>`). 
    - The Debezium Engine Auto-Configuration offers hooks to set the `OffsetCommitPolicy`, `ConnectorCallback`, `CompletionCallback` and `Clock`.  Sensible defaults are implemented.
 - Removed MetadataStore as option for commit offset storage.
 - Removed the spring property shortcuts. Only the native Debezium properties are supported. Use the `debezium.properties.*` prefix for set any native property.
 - Removed the old Debezium suppler&source code. Rename the source to `debezium-source`.

artembilan commented on functions/supplier/pom.xml:
```diff
@@ -27,6 +27,7 @@
         <module>s3-supplier</module>
         <module>twitter-supplier</module>
         <module>cdc-debezium-supplier</module>
+        <module>cdc-debezium-supplier2</module>
```
I don't think we have to do this. 
We are in a new major `4.0` version, so it is fully OK to remove the old code and functionality and apply a new one.
Or you did it deliberately to simplify a review and we will have a follow up PR to clean things up?

tzolov replied:
If the consensus is to remove the old code (i'm strongly for it) I'll remove the `cdc-debezium-common` , `cdc-debezium-boot-starter`, `cdc-debezium-supplier` and `cdc-debezium-source`  projects and 
will make the new `cdc-debezium-supplie2` and `cdc-debezium-source2`  into `cdc-debezium-supplier` and `cdc-debezium-source` instead.

tzolov replied:
resolved

artembilan commented on functions/supplier/cdc-debezium-supplier2/src/test/resources/application.properties:
```diff
@@ -0,0 +1 @@
+spring.autoconfigure.exclude=org.springframework.boot.autoconfigure.mongo.MongoAutoConfiguration
```
I don't think we need this since we do that in the code via `@EnableAutoConfiguration`

tzolov replied:
resolved

artembilan commented on applications/source/cdc-debezium-source2/src/test/docker/mongodb/import-data.sh:
```diff
@@ -0,0 +1,5 @@
+sleep 10s
+
+#populate the db
+/usr/local/bin/init-inventory.sh
```
Doesn't look like this file is in use.

tzolov replied:
It is used in the README. 
Unlike the mysql or postgres, last time i tried I could not automate the mongodb testing as it requires manual interventions. 
This might have changed but i have very low priority in testing mongo stuff.

artembilan commented on applications/source/cdc-debezium-source2/src/test/java/org/springframework/cloud/stream/app/source/cdc/CdcMySqlTestSupport.java:
```diff
@@ -0,0 +1,63 @@
+				"com.mysql.cj.jdbc.Driver",
+				"jdbc:mysql://localhost:" + MAPPED_PORT + "/" + DATABASE_NAME + "?enabledTLSProtocols=TLSv1.2",
+				"root",
+				"debezium");
```
I think all of these can map to auto-configuration properties.
The `@DynamicPropertySource` is a good way to do that from Testcontainers data.

tzolov replied:
Not sure if the @DynamicPropertySource would be useful here as we create the container once per test. 
IMO, this feature would be useful if the same container is shared amongst multiple tests?

artembilan replied:
Not sure what you mean with "once per test" because I see `static GenericContainer debeziumMySQL`.
So, it is really created for all the tests inheriting `CdcMySqlTestSupport`.
That `@DynamicPropertySource` can be used from a `static` initializer as well, e.g.:
```
	@DynamicPropertySource
	static void mongoDbProperties(DynamicPropertyRegistry registry) {
		registry.add("spring.data.mongodb.uri", MongoDbTestContainerSupport::mongoDbUri);
		registry.add("spring.data.mongodb.database", () -> "test");
	}
```

but I still see your point about responsibility concern: this class is not Spring-aware, so probably it is OK to leave as is.

artembilan commented on functions/supplier/cdc-debezium-supplier2/pom.xml:
```diff
@@ -0,0 +1,250 @@
+        <java.version>17</java.version>
+        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+        <version.debezium>2.1.4.Final</version.debezium>
+        <spring-cloud.version>2022.0.2</spring-cloud.version>
```
Why is that?
I thought that Spring Cloud is a parent some way for all of our Stream applications...

tzolov replied:
This is correct. All the java.version, project.build.sourceEncoding and the spring-cloud.version are already defined in the parent stream-applications-build. 
Will remove the duplications.

tzolov replied:
resolved

artembilan commented on functions/supplier/cdc-debezium-supplier2/src/main/java/org/springframework/cloud/fn/supplier/cdc/CdcConfiguration.java:
```diff
@@ -0,0 +1,225 @@
+	public EmbeddedEngineExecutorService embeddedEngine(DebeziumEngine<?> debeziumEngine) {
+
+		return new EmbeddedEngineExecutorService(debeziumEngine) {
+			@PostConstruct
```
This is not OK to perform some low-level resource connection in the initialization phase.
Looks like this `EmbeddedEngineExecutorService` is our class, so let's consider to implement a `SmartLifecycle` contract instead!

tzolov replied:
you're right. This ugly. I will implement the smat lifecycle.

tzolov replied:
resolved

corneil commented:
> * Add support for DB2, Cassandra, Vitess, Spanner
> * Add AVRO format support. Defaults to JSON
> * Remove the Metadata repo configuration.
> * Remove the Spring property sortcuts. Only the native Debezium properties are used.
> * change the native properties prefix from cdc.config to cdc.debezium.
> * Removes the need to fork and modify the Debezium engine code.
> * Add binding name strategy.
> * Update to latest 2.1.4 Final version.
> 
> This is a new Debezium source implementation, build on the Debezium Engine API. I've removed the support for metadata registry as it requires forking and modifying Debezium code. Above changes also means that the new implementation is not backward compatible. This is the reason to add new cdc-debezium-sourc2 and cdc-debezium-spupplier2 instead of modifying the original.

The current implementation is a source and does not implement the contract of functional `Supplier`.
Many of the source implementations provides Spring Integration MessageProducer which is then used to create  `Supplier<Flux<Message<?>>>`

tzolov commented:
@corneil the code and description are updated. The implementation does implement the Supplier contract and exposes Supplier<Flux<Message<?>>>.

artembilan commented on functions/supplier/debezium-supplier/src/main/resources/application.properties:
```diff
@@ -0,0 +1 @@
+spring.cloud.stream.kafka.default.producer.messageKeyExpression=headers['debezium_key'].bytes
```
This one has to go away from the `supplier` module.

tzolov replied:
sure

artembilan commented on functions/supplier/debezium-supplier/pom.xml:
```diff
@@ -0,0 +1,182 @@
+    <description>Debezium Supplier</description>
+
+    <properties>
+        <version.debezium>2.2.0.CR1</version.debezium>
```
There is already a GA for this version.

tzolov replied:
2.2.0.Final is latest GA

artembilan commented on functions/supplier/debezium-supplier/src/main/java/org/springframework/cloud/fn/supplier/debezium/DebeziumEngineAutoConfiguration.java:
```diff
@@ -0,0 +1,198 @@
+	private static final Log logger = LogFactory.getLog(DebeziumEngineAutoConfiguration.class);
+
+	@Bean
+	public Properties debeziumConfiguration(DebeziumProperties properties) {
```
Spring Boot dictates a recommendation to not have a `Properties` as a bean.

Consider to have an ctor injection for the `DebeziumProperties` and create an internal instance for this `Properties` to use in bean method below.

tzolov replied:
Replaced the bean with on-the-fly propety method: getDebeziumNativeConfiguration

artembilan commented on functions/supplier/debezium-supplier/src/main/java/org/springframework/cloud/fn/supplier/debezium/DebeziumProperties.java:
```diff
@@ -0,0 +1,134 @@
+	 * Spring pass-trough wrapper for debezium configuration properties. All properties with a 'debezium.inner.*' prefix
+	 * are native Debezium properties.
+	 */
+	private Map<String, String> inner = defaultConfig();
```
In other cases we just use a name `properties`.
See `KafkaProperties.properties` for example.

tzolov replied:
It is important to streass that those are native Debezium configuration vs. Spring properitieds. 
'Inner' better conveys this meaning.

artembilan reviewed (changes requested):
Looks like there is more fresh MongoDb Docker image for Debezium sample: https://hub.docker.com/r/debezium/example-mongodb/tags.

There is also probably need to add a `StepVerifier` test for our `Supplier<Flux<Message<?>>> debeziumSupplier` bean.

tzolov commented on functions/supplier/debezium-supplier/src/main/java/org/springframework/cloud/fn/supplier/debezium/DebeziumEngineAutoConfiguration.java:
```diff
@@ -0,0 +1,198 @@
+ * <p>
+ * Engine Is designed to be submitted to an {@link Executor} or {@link ExecutorService} for execution by a single
+ * thread, and a running connector can be stopped either by calling {@link #stop()} from another thread or by
+ * interrupting the running thread (e.g., as is the case with {@link ExecutorService#shutdownNow()}).
```
resolved

artembilan reviewed (changes requested):
/home/runner/work/stream-applications/stream-applications/functions/supplier/debezium-supplier/src/test/java/org/springframework/cloud/fn/supplier/debezium/SimpleTest.java:27: Line has leading space characters; indentation should be performed with tabs only. [RegexpSinglelineJava]

tzolov commented on functions/supplier/debezium-supplier/src/main/java/org/springframework/cloud/fn/supplier/debezium/DebeziumEngineAutoConfiguration.java:
```diff
@@ -0,0 +1,184 @@
+				.using(connectorCallback)
+				.using((offsetCommitPolicy != NULL_OFFSET_COMMIT_POLICY) ? offsetCommitPolicy : null)
+				.notifying(changeEventConsumer)
+				.build();
```
While initializing the DebeziumEginine we pass the native properties. E.g. `.using(properties.getDebeziumNativeConfiguration())`

tzolov commented on functions/supplier/debezium-supplier/src/main/java/org/springframework/cloud/fn/supplier/debezium/DebeziumEngineAutoConfiguration.java:
```diff
@@ -0,0 +1,184 @@
+				.notifying(changeEventConsumer)
+				.build();
+
+		logger.debug("Debezium Engine created!");
```
makes sence

tzolov commented on applications/source/debezium-source/README.adoc:
```diff
@@ -0,0 +1,343 @@
+|===
+| Connector | Connector properties
+
+|https://debezium.io/documentation/reference/2.1/connectors/mysql.html[MySQL]
```
yep, 2.2.0.Final is the latest GA

tzolov commented on functions/supplier/debezium-supplier/src/test/java/org/springframework/cloud/fn/supplier/debezium/custom/EmbeddedEngineExecutorService.java:
```diff
@@ -14,39 +14,62 @@
  * @author Christian Tzolov
  */
-public class EmbeddedEngineExecutorService implements Closeable {
+public class EmbeddedEngineExecutorService implements SmartLifecycle, AutoCloseable {
```
tests

tzolov replied:
The `DebeziumEngineAutoConfiguration` is agnostic to the Consumer implementation and can be used with any `Consumer<ChangeEvent>` implementation as shown in the `DebeziumCustomConsumerTest` that internally uses the EmbeddedEngineExecutorService.
The `DebeziumReactiveConsumerConfiguration` provides a specific `Flex` `Consumer<ChangeEvent>` implementation used (implicitly) by the `debezium-source` binders.

tzolov commented on functions/supplier/debezium-supplier/pom.xml:
```diff
@@ -0,0 +1,187 @@
+                </exclusion>
+                <exclusion>
+                    <artifactId>slf4j-log4j12</artifactId>
+                    <groupId>org.slf4j</groupId>
```
Historically, it is messing with Spring's own logging configurations lading to no logs being shown at all. It is safe to keep debezium's internal logging and monitoring away.

tzolov commented:
Addressed all open review recommendations, including a `StepVerifier` test for the `Supplier<Flux<Message<?>>> debeziumSupplier` and updated mongodb testing, though later still require manual testing (description provided in the README).

tzolov commented on functions/supplier/debezium-supplier/src/test/java/org/springframework/cloud/fn/supplier/debezium/custom/DebeziumEngineAutoConfigurationTestTest.java:
```diff
@@ -0,0 +1,179 @@
+	}
+
+	@SpringBootConfiguration
+	@EnableAutoConfiguration(exclude = { DataSourceAutoConfiguration.class })
```
Yes it looks like. The test pass OK, but i can see mongo related exceptions in the log.

tzolov commented on functions/supplier/debezium-supplier/pom.xml:
```diff
@@ -0,0 +1,187 @@
+
+        <dependency>
+            <groupId>io.projectreactor</groupId>
+            <artifactId>reactor-test</artifactId>
```
who am I to argue with your IDE ;)

tzolov commented on applications/source/debezium-source/README.adoc:
```diff
@@ -0,0 +1,343 @@
+
+==== DB2, Vitess, Spanner
+
+samples coming soon ...
```
sure

tzolov commented on applications/source/debezium-source/README.adoc:
```diff
@@ -0,0 +1,343 @@
+|===
+| Connector | Connector properties
+
+|https://debezium.io/documentation/reference/2.2/connectors/mysql.html[MySQL]
```
Not sure what you mean? the https://debezium.io/documentation/reference/2.2/connectors/mysql.html is opening ok and corresponds to the connector version we support.

artembilan replied:
Right. I just want to avoid burden of supporting updates to this doc when we just would move to Debezium `2.3`, for example.
Why just not have a "versionless" link in the doc while on upgrade we would mostly just change the version in the POM?

artembilan reviewed (approved):
LGTM. Merging...

Anything missed can be addressed in other contributions.

artembilan commented:
Merged as https://github.com/spring-cloud/stream-applications/commit/6cf66ec779faaed28a759dde409b034c92cb7356 after rebasing.

@tzolov , huge thanks: as always tremendous work!

Looking forward for more! 😄
